cmd/maya-apiserver/app/server/backup_endpoint_v1alpha1.go

```diff
@@ -0,0 +1,388 @@
+	"github.com/openebs/maya/pkg/client/generated/clientset/internalclientset"
+	snapshot "github.com/openebs/maya/pkg/snapshot/v1alpha1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+)
+
+type backupAPIOps struct {
+	req  *http.Request
+	resp http.ResponseWriter
+}
+
+// backupV1alpha1SpecificRequest deals with backup API requests
+func (s *HTTPServer) backupV1alpha1SpecificRequest(resp http.ResponseWriter, req *http.Request) (interface{}, error) {
+	backupOp := &backupAPIOps{
+		req:  req,
+		resp: resp,
+	}
+
+	switch req.Method {
+	case "POST":
+		return backupOp.create()
+	case "GET":
+		return backupOp.get()
+	}
+	return nil, CodedError(405, ErrInvalidMethod)
+}
+
+// Create is http handler which handles backup create request
+func (bOps *backupAPIOps) create() (interface{}, error) {
+	bkp := &v1alpha1.BackupCStor{}
+
+	err := decodeBody(bOps.req, bkp)
+	if err != nil {
+		return nil, err
+	}
+
+	// backup name is expected
+	if len(strings.TrimSpace(bkp.Spec.BackupName)) == 0 {
+		return nil, CodedError(400, fmt.Sprintf("Failed to create backup: missing backup name "))
+	}
+
+	// namespace is expected
+	if len(strings.TrimSpace(bkp.Namespace)) == 0 {
+		return nil, CodedError(400, fmt.Sprintf("Failed to create backup '%v': missing namespace", bkp.Spec.BackupName))
+	}
+
+	// volume name is expected
+	if len(strings.TrimSpace(bkp.Spec.VolumeName)) == 0 {
+		return nil, CodedError(400, fmt.Sprintf("Failed to create backup '%v': missing volume name", bkp.Spec.BackupName))
+	}
+
+	// backupIP is expected
+	if len(strings.TrimSpace(bkp.Spec.BackupDest)) == 0 {
+		return nil, CodedError(400, fmt.Sprintf("Failed to create backup '%v': missing backupIP", bkp.Spec.BackupName))
+	}
+
+	// snapshot name is expected
+	if len(strings.TrimSpace(bkp.Spec.SnapName)) == 0 {
+		return nil, CodedError(400, fmt.Sprintf("Failed to create backup '%v': missing snapName", bkp.Spec.BackupName))
+	}
+
+	openebsClient, _, err := loadClientFromServiceAccount()
+	if err != nil {
+		return nil, CodedError(400, fmt.Sprintf("Failed to create openEBSClient '%v'", err))
+	}
+
+	if err = createSnapshotForBackup(bkp); err != nil {
+		return nil, CodedError(400, fmt.Sprintf("Failed to create snapshot '%v'", err))
+	}
+
+	bkp.Name = bkp.Spec.SnapName + "-" + bkp.Spec.VolumeName
+
+	// find healthy CVR
+	cvr := findHealthyCVR(openebsClient, bkp.Spec.VolumeName)
+	if cvr == nil {
+		return nil, CodedError(400, fmt.Sprintf("Failed to find healthy replica"))
+	}
+
+	bkp.ObjectMeta.Labels = map[string]string{
+		"cstorpool.openebs.io/uid":     cvr.ObjectMeta.Labels["cstorpool.openebs.io/uid"],
+		"openebs.io/persistent-volume": cvr.ObjectMeta.Labels["openebs.io/persistent-volume"],
+		"openebs.io/backup":            bkp.Spec.BackupName,
+	}
+
+	// Find last backup snapshot name
+	lastsnap, err := getLastBackupSnap(openebsClient, bkp)
+	if err != nil {
+		return nil, CodedError(400, fmt.Sprintf("Failed create lastbackup"))
```

Comment:
add more info to the error message itself for better debuggability.
For ex., here, it makes sense to log backupname/volnamet with schedule name. Same comment applies at many other places.

Reply:
ok